pkg/virt-controller/watch/migration.go

```diff
@@ -1533,19 +1533,28 @@ func (c *MigrationController) alertIfHostModelIsUnschedulable(vmi *virtv1.Virtua
 	}
 }
 
-func prepareNodeSelectorForHostCpuModel(node *k8sv1.Node, pod *k8sv1.Pod) error {
+func prepareNodeSelectorForHostCpuModel(node *k8sv1.Node, pod *k8sv1.Pod, vmi *virtv1.VirtualMachineInstance, sourcePod *k8sv1.Pod) error {
 	var hostCpuModel, hostCpuModelLabelKey, hostModelLabelValue string
 
 	for key, value := range node.Labels {
 		if strings.HasPrefix(key, virtv1.HostModelCPULabel) {
 			hostCpuModel = strings.TrimPrefix(key, virtv1.HostModelCPULabel)
 			hostModelLabelValue = value
 		}
-
-		if strings.HasPrefix(key, virtv1.HostModelRequiredFeaturesLabel) {
+		if vmi.Status.MigratedAtLeastOnce == false && strings.HasPrefix(key, virtv1.HostModelRequiredFeaturesLabel) {
```

Comment:
Perhaps we don't need `vmi.Status.MigratedAtLeastOnce` field at all?
I think we can simply rely on the presence of the previous node selector and just copy it to the destination pod.

Reply:
The key here is that previously we were searching by the `HostModelCPULabel` prefix, but we should have been searching by the `CPUModelLabel` prefix

So this
```
hostCpuModelLabelKey = virtv1.HostModelCPULabel + hostCpuModel
pod.Spec.NodeSelector[hostCpuModelLabelKey] = hostModelLabelValue
```
 would become 

```
 cpuModelLabelKey = virtv1.CPUModelLabel + hostCpuModel                                                 
pod.Spec.NodeSelector[cpuModelLabelKey] = hostModelLabelValue     
```

Reply:
Hey Vladik,
Thanks for the comments.
your idea sounds good but what if the initial nodes doesn't have any required features labels
In that case if we won't use `vmi.Status.MigratedAtLeastOnce` we will be coupled to the target node selectors all the way.
For instance:
we start the vmi on node01 that doesn't have any required features level and than migrate to node02 that has required feature A

Reply:
Hey Vladik,
Thanks for reviewing.
Your idea sounds good but what if the initial node doesn't have any required features label i think that in that case if we won't use `vmi.Status.MigratedAtLeastOnce` than we will be coupled to the target node required feature.
For Instance:
If we won't use `vmi.Status.MigratedAtLeastOnce` and than vmi with host-model is scheduled to the initial node node01 without any required feature and than migrate to node02 with required feature A.
in that case when we will try to migrate the vmi again from node02 we will ask for node with required feature A even though we do not need it.


and about the `cpuModelLabelKey` i have a similar PR to fix this issue, i'm waiting for this one to be approved to publish it because i use the test's Context to test this fix and i prefer to keep this PR small and focused on only one issue to make it easier to review in the future.
just to be clear the next PR is to solve this issue:
`vmi with host-model can't migrate if the target node does't have the same host model even if the target node support the host-model of the source node and has all the required features.`

Reply:
> Your idea sounds good but what if the initial node doesn't have any required feature labels i think that in that case if we won't use vmi.Status.MigratedAtLeastOnce than we will be coupled to the target node required features.

If the problem we're trying to solve here involves knowledge of the first node a VMI runs on, can we store all that information in the VMI status after the VMI starts so we don't have to keep track of whether migrations have already occurred or not?

Reply:
> Hey Vladik, Thanks for reviewing. Your idea sounds good but what if the initial node doesn't have any required feature labels i think that in that case if we won't use `vmi.Status.MigratedAtLeastOnce` than we will be coupled to the target node required features. On second thought we can search for `CPUModelLabel` selector instead of `CPUFeatureLabel` that is always in virt-launcher after the first migration to solve this, i will update the PR soon.

The function I've posted above searches for either `CPUFeatureLabel` or `CPUModelLabel` if one of these labels exists in the source pod nodeSelector then we should propagate this to the destination pod nodeSelector. 
Otherwise, we will create a new nodeSelector for the destination pod that is based on the translation of the host-model CPU, as this is probably the first time we are migrating this VMI.

```
        for key, value := range sourcePod.Spec.NodeSelector {                                                          
                if strings.Contains(key, virtv1.CPUFeatureLabel) || strings.Contains(key, virtv1.CPUModelLabel) {      
                        pod.Spec.NodeSelector[key] = value
                        haveCPUNodeSelector := true
                }               
        }               
        if !haveCPUNodeSelector {         
```


> 
> and about the `cpuModelLabelKey` i have a similar branch to solve this issue, i'm waiting for this one to be merged to publish it because i use this new test's Context to test this issue modifications and i prefer to keep this PR small and more cohesive to make it easier to review in the future. just to be clear the next PR is to solve this issue: `vmi with host-model can't migrate if the target node does't have the same host model even if the target node support the host-model of the source node and has all the required features.` and it will be publish immediately after this one will get merged

In my view, these two changes should happen in the same PR. 
Previously, we were generating the nodeSelector using the `HostModelCPULabel` field. Without changing the cpuModelLabelKey, I don't see how can we test this PR.

Does it make sense?

Reply:
@davidvossel we probably could, but perhaps we can get away without updating the vmi status.
On the first migration, the translated CPU model will be stored in the destination pod nodeSelector.
Any following migration could simply reuse this nodeSelector.

My concern with storing this information in the VMI status is that it won't be backward compatible.

Reply:
```
The function I've posted above searches for either CPUFeatureLabel or CPUModelLabel if one of these labels exists in the source pod nodeSelector then we should propagate this to the destination pod nodeSelector.
Otherwise, we will create a new nodeSelector for the destination pod that is based on the translation of the host-model CPU, as this is probably the first time we are migrating this VMI.
```
Right, I missed the  `OR operator || ` sorry about that.

Reply:
@vladikr i updated the PR to not include the new field in vmi.status, updated the description of this PR.

```
In my view, these two changes should happen in the same PR.
Previously, we were generating the nodeSelector using the HostModelCPULabel field. Without changing the cpuModelLabelKey, I don't see how can we test this PR.
```
I also added the modification of my second branch to this PR including a test that disable the target node labeller and simulate a target node with fake host-model that support the source node host-model.
(if it is possible to simulate it in the cluster otherwise i skip the test just like the first test)

regarding the changes to the `hostCpuModelLabelKey` :
This modification is necessary but we also need to include the host-model cpuModel of a node as part of the supported cpuModels in the node(using `CPUModelLabel + <host-model> : true` ) , i included this modification as part of the last commit of this PR.
The test in the third commit of this PR doesn't pass without both modification.

Honestly i think that keeping PR as cohesive as possible is best practice.
From my point of view we should deal with one bug in each PR and add a test for it to make sure it doesn't happen again,
this way it is easier to follow and understand in future reviews.